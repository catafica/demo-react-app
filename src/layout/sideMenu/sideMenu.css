
.sideNav {
  height: 100%; /* Full-height: remove this if you want "auto" height */
  width: 240px; /* Set the width of the sidebar */
  position: fixed; /* Fixed Sidebar (stay in place on scroll) */
  z-index: 1; /* Stay on top */
  top: 0; /* Stay at the top */
  left: 0;
  background-color: white;/* Black */
  overflow-x: hidden; /* Disable horizontal scroll */
  padding-top: 20px;
  border-right:2px solid #11BCD0;
  /* background: -webkit-linear-gradient(#E5E6E8, white); */


}
.img-responsive.myLogo{
  width:200px;
  opacity:0;
}

.form-group{
  margin-top:5px;
  width:72%;
  margin-left:auto;
  margin-right:auto;
  opacity:0;
}



.form-group label{
  text-align:left;
  font-size:20px;  
}

/* The navigation menu links */
.sideNav a {
  padding: 6px 8px 6px 16px;
  text-decoration: none;
  font-size: 20px;
  display: block;
  transition: color 0.5s;
}

/* When you mouse over the navigation links, change their color */
.sideNav a:hover {
  color: #FF5E80;
}

::-webkit-scrollbar {
    width: 0px;  /* remove scrollbar space */
    background: transparent;  /* optional: just make scrollbar invisible */
}
/* optional: show position indicator in red */


i{
  margin-top:10px;
  font-size:25px;
  color: #FF5E80;
}

.animated{
  animation: jackInTheBox 1.1s ease-out forwards;
}


@keyframes jackInTheBox {
  0% {
    opacity: 0;
    transform: scale(0.1) rotate(30deg);
    transform-origin: center bottom;
  }

  50% {
    transform: rotate(-10deg);
  }

  70% {
    transform: rotate(3deg);
  }

  100% {
    opacity: 1;
    transform: scale(1);
  }
}
