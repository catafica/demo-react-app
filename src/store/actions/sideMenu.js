import * as actionTypes from './actionTypes';

export const addDog = (name, activation) => {
  return {
    type: actionTypes.ADD_DOG,
    value: name,
    activation: activation
  };
};

export  const activate = (id) => {
  return {
    type: actionTypes.ACTIVATE,
    id: id
  };
};

export const initDogs = () => {
  return {
    type: actionTypes.INIT_DOGS
  }
}

export const addDogList = (list) => {
  return {
    type: actionTypes.ADD_DOG_LIST,
    list
  }
}
