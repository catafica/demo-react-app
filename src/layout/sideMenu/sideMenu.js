import React, { Component } from 'react';
import DogBreed from './dogBreed/dogBreed';
import dogsLogo from '../../assets/images/downtown-dogs4.jpg';
import './sideMenu.css';
import * as sideMenuActions from '../../store/actions/sideMenu';
import { connect } from 'react-redux';


export class SideMenu extends Component {


 state = {
   dogLogoAppear :false,
   dogBreeds: [],
   search:''
 }

 componentDidMount = () => {
   this.props.initiateDogs();
   this.makeLogoAppear();
   this.addDogBreeds();
 }

 makeLogoAppear = () => {
   setTimeout( () => {
     this.setState({dogLogoAppear: true});
   },800)
 }

 addDogBreeds = () => {
  setTimeout( () => {
    this.setState({dogBreeds: this.props.dogList});
   }, 800)
 }


  render = () => {

    let formGroup = 'form-group';
    let imgClassList = 'img-responsive myLogo';
    if (this.state.dogLogoAppear){
      imgClassList = 'img-responsive myLogo animated';
      formGroup = 'form-group animated';
    }

    return (
      <div className="sideNav">
          <a style={{marginTop:'-10px'}}>
            <img className={imgClassList} src={dogsLogo}></img>
          </a>
          <div className={formGroup}>
            <label>Choose dog type:</label>
            <input
             onChange={event => this.setState({search: event.target.value})}
             type="text"
             id="dogType"
             className="form-control"
             name="dogType"
             placeholder="exp. akita"
             >
             </input>
          </div>
          <div style={{height:'65vh', width:'100%', overflow:'scroll', overflowX:'hidden'}}>
          {
            this.state.dogBreeds.map( (dog, i)=> {
            const {search} = this.state;
            if ( search !== '' && dog.name.toLowerCase().indexOf(search.toLowerCase()) === -1){
              return null;
            }

            return (
               <DogBreed
                  key={i}
                  name={dog.name}
                  id={dog.id}
                  activated={dog.activation}
                  clicked={() => this.props.activate(dog.id)}
                />
              )
          })
          }
          </div>
          <div>
            <i className="glyphicon glyphicon-menu-down"></i>
          </div>
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    dogTypes: state.dogTypes,
    dogList: state.dogList
  };
};

const mapDispatchToProps = dispatch => {
  return{
    onAddDog: (name, activation) => dispatch(sideMenuActions.addDog(name,activation)),
    activate: (id) => dispatch(sideMenuActions.activate(id)),
    initiateDogs: () => dispatch(sideMenuActions.initDogs())
  };
};



export default connect(mapStateToProps, mapDispatchToProps)(SideMenu);
