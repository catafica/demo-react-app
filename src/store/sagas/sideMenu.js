import axios from 'axios';
import { put } from 'redux-saga/effects';
import * as sideMenuActions from '../actions/sideMenu';
import { delay } from 'redux-saga';


export function* initDogsSaga(action){
  try {
    const response = yield axios.get('https://dog.ceo/api/breeds/list/all');
    const list = [];
    for (let key in response.data.message){
      yield put(sideMenuActions.addDog(key, false));
      list.push({name: key, id:new Date(), activation: false});
    }
    yield put(sideMenuActions.addDogList(list));
  } catch (e){

  }

}



//
// 'https://dog.ceo/api/breed/' + this.props.currentDogName +'/images'
