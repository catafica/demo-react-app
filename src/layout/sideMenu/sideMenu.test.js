import React from 'react';
import { shallow } from 'enzyme';
import {SideMenu} from './sideMenu';



describe('SideMenu', () => {
  const id = new Date();
  const mockInitiateDogs = jest.fn();
  const mockDogList = jest.fn();
  const mockActivate = jest.fn();
  const props = {initiateDogs: mockInitiateDogs, dogList: mockDogList, clicked:mockActivate};
  const sideMenu = shallow(<SideMenu {...props}/>);

  it('renders properly', () => {
    expect(sideMenu).toMatchSnapshot();
  })

  it('initializes the `dogLogoAppear` property in the state', () => {
    expect(sideMenu.state().dogLogoAppear).toEqual(false);
  })

  describe('after `componentDidMount()` was called', () => {
    beforeEach(() => {
        sideMenu.instance().componentDidMount();
      });

    it('expects to `initiateDogs` prop to be called', () => {
      expect(mockInitiateDogs).toHaveBeenCalled();
    })

    describe('after the callback makeLogoAppear() has been called', () =>{
      beforeEach( () => {
        sideMenu.instance().makeLogoAppear()
      })

      it('changes the dogLogoAppear property in the state after a few miliseconds', () => {
        setTimeout( () => {
          expect(sideMenu.state().dogLogoAppear).toEqual(true);
        },801)

        })
      })
      describe('after the callback addDogBreeds() has been called', () =>{
        beforeEach( () => {
          sideMenu.instance().addDogBreeds()
        })

        it('updated the state with the dogBreeds list', () => {
          expect(sideMenu.state().dogBreeds).not.toBe([]);
        })
      })

      it('creates a DogBreed component after a few miliseconds', () => {
        setTimeout( () => {
          expect(sideMenu.find('DogBreed').exists()).toBe(true);
        },801)
      })

      describe('when typing into the `choose dog type` input', () =>{
        const dogType = 'Akita';

        beforeEach( () => {
          sideMenu.find('.form-control').simulate('change', {target: { value: dogType }});
        })

        it('updates the search in `state`', () => {
          expect(sideMenu.state().search).toEqual(dogType);
        })
      })  
      describe('when user click one of the dog breeds that appeared in the side menu', () => {
        beforeEach( () => {
          setTimeout( () => {
            sideMenu.find('DogBreed').simulate('click');
          },801)

        })

        it('expects to fire `props.activate` which dispatches the ACTIVATE action', () => {
            setTimeout( () => {
              expect(mockActivate).toHaveBeenCalledWith(id);
            }, 801)
        })
      })
    })


})
