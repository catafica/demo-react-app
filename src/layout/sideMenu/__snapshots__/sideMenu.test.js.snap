// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideMenu renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SideMenu
    clicked={[MockFunction]}
    dogList={[MockFunction]}
    initiateDogs={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <a
          style={
            Object {
              "marginTop": "-10px",
            }
          }
        >
          <img
            className="img-responsive myLogo"
            src="downtown-dogs4.jpg"
          />
        </a>,
        <div
          className="form-group"
        >
          <label>
            Choose dog type:
          </label>
          <input
            className="form-control"
            id="dogType"
            name="dogType"
            onChange={[Function]}
            placeholder="exp. akita"
            type="text"
          />
        </div>,
        <div
          style={
            Object {
              "height": "65vh",
              "overflow": "scroll",
              "overflowX": "hidden",
              "width": "100%",
            }
          }
        />,
        <div>
          <i
            className="glyphicon glyphicon-menu-down"
          />
        </div>,
      ],
      "className": "sideNav",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <img
            className="img-responsive myLogo"
            src="downtown-dogs4.jpg"
          />,
          "style": Object {
            "marginTop": "-10px",
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "img-responsive myLogo",
            "src": "downtown-dogs4.jpg",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        "type": "a",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label>
              Choose dog type:
            </label>,
            <input
              className="form-control"
              id="dogType"
              name="dogType"
              onChange={[Function]}
              placeholder="exp. akita"
              type="text"
            />,
          ],
          "className": "form-group",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Choose dog type:",
            },
            "ref": null,
            "rendered": "Choose dog type:",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-control",
              "id": "dogType",
              "name": "dogType",
              "onChange": [Function],
              "placeholder": "exp. akita",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "style": Object {
            "height": "65vh",
            "overflow": "scroll",
            "overflowX": "hidden",
            "width": "100%",
          },
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <i
            className="glyphicon glyphicon-menu-down"
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "glyphicon glyphicon-menu-down",
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <a
            style={
              Object {
                "marginTop": "-10px",
              }
            }
          >
            <img
              className="img-responsive myLogo"
              src="downtown-dogs4.jpg"
            />
          </a>,
          <div
            className="form-group"
          >
            <label>
              Choose dog type:
            </label>
            <input
              className="form-control"
              id="dogType"
              name="dogType"
              onChange={[Function]}
              placeholder="exp. akita"
              type="text"
            />
          </div>,
          <div
            style={
              Object {
                "height": "65vh",
                "overflow": "scroll",
                "overflowX": "hidden",
                "width": "100%",
              }
            }
          />,
          <div>
            <i
              className="glyphicon glyphicon-menu-down"
            />
          </div>,
        ],
        "className": "sideNav",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <img
              className="img-responsive myLogo"
              src="downtown-dogs4.jpg"
            />,
            "style": Object {
              "marginTop": "-10px",
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "img-responsive myLogo",
              "src": "downtown-dogs4.jpg",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "a",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label>
                Choose dog type:
              </label>,
              <input
                className="form-control"
                id="dogType"
                name="dogType"
                onChange={[Function]}
                placeholder="exp. akita"
                type="text"
              />,
            ],
            "className": "form-group",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Choose dog type:",
              },
              "ref": null,
              "rendered": "Choose dog type:",
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-control",
                "id": "dogType",
                "name": "dogType",
                "onChange": [Function],
                "placeholder": "exp. akita",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "style": Object {
              "height": "65vh",
              "overflow": "scroll",
              "overflowX": "hidden",
              "width": "100%",
            },
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <i
              className="glyphicon glyphicon-menu-down"
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "glyphicon glyphicon-menu-down",
            },
            "ref": null,
            "rendered": null,
            "type": "i",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
